### PR TITLE
Most of the matchers on String really should be matchers on CharSeque…

### LINKED ISSUE #2224: Most of the matchers on String really should be matchers on CharSequence
Matchers on String are really too specific in that they only match on String, but CharSequence is the interface that String implements and most of the matchers don't care about the difference.

### PR DESCRIPTION
**Pull request by sksamuel**
Closes #2224

### CODE REVIEW

File: kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/string/empty.kt [hunk 1]
```diff
@@ -0,0 +1,44 @@
+package io.kotest.matchers.string
+
+import io.kotest.assertions.show.show
+import io.kotest.matchers.MatcherResult
+import io.kotest.matchers.neverNullMatcher
+import io.kotest.matchers.should
+import io.kotest.matchers.shouldNot
+
+fun <A : CharSequence> A?.shouldBeEmpty(): A? {
+   this.toString() should beEmpty()
```
Comment (myuwono) [inline kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/string/empty.kt:10]: hmm `null.toString()` will yield the string "null"

Comment (sksamuel) [reply kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/string/empty.kt:10]: Good spot.

File: kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/string/case.kt [hunk 2]
```diff
@@ -6,13 +6,13 @@ import io.kotest.matchers.neverNullMatcher
 import io.kotest.matchers.should
 import io.kotest.matchers.shouldNot
 
-fun String?.shouldBeUpperCase(): String? {
-   this should beUpperCase()
+fun <A : CharSequence> A?.shouldBeUpperCase(): A? {
+   this.toString() should beUpperCase()
```
Comment (myuwono) [inline kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/string/case.kt:10]: `null.toString()`

File: kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/string/ends.kt [hunk 3]
```diff
@@ -0,0 +1,14 @@
+package io.kotest.matchers.string
+
+import io.kotest.matchers.should
+import io.kotest.matchers.shouldNot
+
+infix fun <A : CharSequence> A?.shouldEndWith(suffix: String): A? {
+   this.toString() should endWith(suffix)
```
Comment (myuwono) [inline kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/string/ends.kt:7]: `null.toString()`

i start to wonder if we should change `endWith` and other `String` matchers to work based on `A : CharSequence` now? i.e.

```kotlin

fun <A : CharSequence> beEmpty() = neverNullMatcher<A> { value ->
   MatcherResult(
      value.toString().isEmpty(),
      "${value.toString().show().value} should be empty",
      "${value.toString().show().value} should not be empty"
   )
}

```

**Comment by sksamuel** [comment]
Closing in favour of #2278

**Pull request closed without merging**